libraries/hotstuff/qc_chain.cpp

```diff
@@ -0,0 +1,1224 @@
+         return nullptr;
+      const hs_proposal_message & proposal = ps_it->second;
+      return &proposal;
+#else
+      proposal_store_type::nth_index<0>::type::iterator itr = _proposal_store.get<by_proposal_id>().find( proposal_id );
+      if (itr == _proposal_store.get<by_proposal_id>().end())
+         return nullptr;
+      return &(*itr);
+#endif
+   }
+
+   bool qc_chain::insert_proposal(const hs_proposal_message & proposal) {
+      std::lock_guard g( _state_mutex );
+#ifdef QC_CHAIN_SIMPLE_PROPOSAL_STORE
+      uint64_t proposal_height = proposal.get_height();
+      ps_height_iterator psh_it = _proposal_stores_by_height.find( proposal_height );
+      if (psh_it == _proposal_stores_by_height.end()) {
+         _proposal_stores_by_height.emplace( proposal_height, proposal_store() );
+         psh_it = _proposal_stores_by_height.find( proposal_height );
+      }
+      proposal_store & pstore = psh_it->second;
+      const fc::sha256 & proposal_id = proposal.proposal_id;
+      ps_iterator ps_it = pstore.find( proposal_id );
+      if (ps_it != pstore.end())
+         return false; // duplicate proposal insertion, so don't change anything actually
+      _proposal_height.emplace( proposal_id, proposal_height );
+      pstore.emplace( proposal_id, proposal );
+      return true;
+#else
+      if (get_proposal( proposal.proposal_id ) != nullptr)
+         return false;
+      _proposal_store.insert(proposal); //new proposal
+      return true;
+#endif
+   }
+
+   void qc_chain::get_state( finalizer_state& fs ) const {
+      std::lock_guard g( _state_mutex );
+      fs.chained_mode           = _chained_mode;
+      fs.b_leaf                 = _b_leaf;
+      fs.b_lock                 = _b_lock;
+      fs.b_exec                 = _b_exec;
+      fs.b_finality_violation   = _b_finality_violation;
+      fs.block_exec             = _block_exec;
+      fs.pending_proposal_block = _pending_proposal_block;
+      fs.v_height               = _v_height;
+      fs.high_qc                = _high_qc;
+      fs.current_qc             = _current_qc;
+      fs.schedule               = _schedule;
+#ifdef QC_CHAIN_SIMPLE_PROPOSAL_STORE
+      ps_height_iterator psh_it = _proposal_stores_by_height.begin();
+      while (psh_it != _proposal_stores_by_height.end()) {
+         proposal_store &pstore = psh_it->second;
+         ps_iterator ps_it = pstore.begin();
+         while (ps_it != pstore.end()) {
+            fs.proposals.insert( *ps_it );
+            ++ps_it;
+         }
+         ++psh_it;
+      }
+#else
+      auto hgt_itr = _proposal_store.get<by_proposal_height>().begin();
+      auto end_itr = _proposal_store.get<by_proposal_height>().end();
+      while (hgt_itr != end_itr) {
+         const hs_proposal_message & p = *hgt_itr;
+         fs.proposals.emplace( p.proposal_id, p );
+         ++hgt_itr;
+      }
+#endif
+   }
+
+   uint32_t qc_chain::positive_bits_count(fc::unsigned_int value){
+      boost::dynamic_bitset b(21, value);
```

Comment:
This assumes 21 producers. Also since `unsigned_int` is used for value this has a upper limit of 32 finalizers (currently producers). Seems like `quorum_certificate.active_finalizers` should be a `std::string` instead to support greater than 32 finalizers. Note this affects on wire format so we need to get it right before 5.0.

Reply:
Not sure what the most appropriate type is for this. String might actually work, if serialization is efficient and compact.

Reply:
String should be fine. `boost::dynamic_bitset` supports string and serialization overhead should be fine I think.